File: src/updater.js
// ================================================================
// src/updater.js
// ----------------------------------------------------------------
// Settings > Updates. Talks to the commands registered in
// src-tauri/src/updater.rs (check_for_updates, install_update_now,
// get_app_version, set_auto_check_updates) through
// window.__TAURI__.core.invoke — the same bridge main.js already uses
// for everything else, since this project loads plain <script> files
// rather than going through a bundler (see the comment above main.js's
// own tauriBridge setup near the top of that file).
//
// The actual check runs on a background thread in Rust (see
// spawn_background_checks in updater.rs), so it keeps happening even
// while this window is closed and only the desktop widget remains open.
// This file just reflects whatever "update-status" event Rust last
// emitted, and lets the user trigger an on-demand check or install.
// ================================================================

(function () {
  const tauriBridge = window.__TAURI__;
  const tauriInvoke = tauriBridge && tauriBridge.core && tauriBridge.core.invoke;
  const tauriListen = tauriBridge && tauriBridge.event && tauriBridge.event.listen;

  const AUTO_CHECK_PREF_KEY = "chrononAutoCheckUpdates";

  const state = {
    status: { state: "idle" },
    version: null,
    lastChecked: null,
    installing: false,
  };

  function readAutoCheckPref() {
    const raw = localStorage.getItem(AUTO_CHECK_PREF_KEY);
    return raw === null ? true : raw === "1";
  }

  function writeAutoCheckPref(enabled) {
    localStorage.setItem(AUTO_CHECK_PREF_KEY, enabled ? "1" : "0");
  }

  function formatLastChecked(date) {
    if (!date) return "";
    const time = date.toLocaleTimeString([], { hour: "2-digit", minute: "2-digit" });
    return `Last checked ${time}`;
  }

  // Maps the current status + a couple of local flags to exactly what
  // the card should show. Kept as one pure function so render() never
  // has to duplicate this branching.
  function view() {
    const runningVersion = state.version ? `v${state.version}` : "";
    switch (state.status.state) {
      case "checking":
        return { dot: "checking", text: "Checking for updates…", btnText: "Checking…", btnDisabled: true };

      case "available":
        return {
          dot: "available",
          text: `Update available — v${state.status.version}`,
          btnText: state.installing ? "Installing…" : "Download & Install",
          btnDisabled: state.installing,
        };

      case "downloading":
        return {
          dot: "downloading",
          text: `Downloading update…${state.status.percent != null ? " " + state.status.percent + "%" : ""}`,
          btnText: "Downloading…",
          btnDisabled: true,
        };

      case "up-to-date":
        return {
          dot: "up-to-date",
          text: runningVersion ? `You're up to date (${runningVersion})` : "You're up to date",
          btnText: "Check for Updates Now",
          btnDisabled: false,
        };

      case "error":
        return {
          dot: "error",
          text: state.status.message || "Couldn't check for updates.",
          btnText: "Try Again",
          btnDisabled: false,
        };

      default:
        return {
          dot: "idle",
          text: runningVersion ? `Running ${runningVersion}` : "",
          btnText: "Check for Updates Now",
          btnDisabled: false,
        };
    }
  }

  function render() {
    const dot = document.getElementById("updateDot");
    const text = document.getElementById("updateStatusText");
    const btn = document.getElementById("updateActionBtn");
    const versionEl = document.getElementById("updateCurrentVersion");
    const lastCheckedEl = document.getElementById("updateLastChecked");
    if (!dot || !text || !btn) return;

    const v = view();
    dot.dataset.state = v.dot;
    text.textContent = v.text;
    btn.textContent = v.btnText;
    btn.disabled = v.btnDisabled;
    if (versionEl) versionEl.textContent = state.version ? `v${state.version}` : "—";
    if (lastCheckedEl) lastCheckedEl.textContent = formatLastChecked(state.lastChecked);
  }

  async function loadVersion() {
    if (!tauriInvoke) return;
    try {
      state.version = await tauriInvoke("get_app_version");
    } catch (err) {
      // Leave state.version null — render() just shows "—" for it.
      console.warn("[Updater] Couldn't read app version:", err);
    }
    render();
  }

  async function onActionClick() {
    if (!tauriInvoke) return;
    const v = view();
    if (v.btnDisabled) return;

    if (state.status.state === "available") {
      state.installing = true;
      render();
      try {
        await tauriInvoke("install_update_now");
        // On Windows the app exits partway through this call to hand
        // off to the installer, so nothing after this line runs in
        // practice on that platform — see the comment above
        // install_update_now in updater.rs.
      } catch (err) {
        state.status = {
          state: "error",
          message: "Couldn't download or install the update. Check your internet connection and try again.",
        };
        state.installing = false;
        render();
      }
      return;
    }

    state.status = { state: "checking" };
    render();
    const result = await tauriInvoke("check_for_updates");
    state.lastChecked = new Date();
    if (result && result.started === false) {
      state.status = { state: "error", message: result.reason || "Couldn't check for updates." };
    }
    // On success, the real outcome (available / up-to-date / error)
    // arrives shortly after via the "update-status" event listener
    // below — run_check() in updater.rs emits it either way.
    render();
  }

  function initToggle() {
    const toggle = document.getElementById("autoCheckUpdates");
    if (!toggle) return;

    const enabled = readAutoCheckPref();
    toggle.checked = enabled;
    if (tauriInvoke) tauriInvoke("set_auto_check_updates", { enabled }).catch(() => {});

    toggle.addEventListener("change", () => {
      writeAutoCheckPref(toggle.checked);
      if (tauriInvoke) tauriInvoke("set_auto_check_updates", { enabled: toggle.checked }).catch(() => {});
    });
  }

  function init() {
    if (!tauriInvoke) return; // e.g. index.html opened directly in a browser while tweaking CSS

    initToggle();
    loadVersion();

    const btn = document.getElementById("updateActionBtn");
    if (btn) btn.addEventListener("click", onActionClick);

    if (tauriListen) {
      tauriListen("update-status", (event) => {
        state.status = event.payload;
        if (state.status.state === "checking") state.lastChecked = new Date();
        if (state.status.state !== "available") state.installing = false;
        render();
      });
    }

    render();
  }

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", init);
  } else {
    init();
  }
})();

File: src-tauri/src/updater.rs
// ================================================================
// src-tauri/src/updater.rs
// ----------------------------------------------------------------
// GitHub-release auto-update, built on tauri-plugin-updater +
// tauri-plugin-process. Plays the same role here that electron-updater
// plays in Project Playnck's main.js: a background check that keeps
// running for as long as the app process is alive (which, for Chronon,
// means as long as the desktop widget window exists — not just while
// the editor/settings window happens to be open), plus a couple of
// commands the Settings > Updates UI calls directly.
//
// Frontend counterpart: src/updater.js.
// ================================================================

use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::{Arc, Mutex};
use std::time::Duration;

use serde::Serialize;
use tauri::{AppHandle, Emitter, Manager, State};
use tauri_plugin_updater::{Update, UpdaterExt};

/// Holds the `Update` handle from the most recent successful check, so
/// `install_update_now` can act on it without checking again first.
/// electron-updater keeps this kind of state internally; the Tauri
/// plugin expects the app to hold onto it explicitly.
#[derive(Default)]
pub struct PendingUpdate(pub Mutex<Option<Update>>);

/// Whether the background loop below should actually check. Mirrors
/// Settings > Updates > "Automatically check for updates" — the
/// frontend syncs its saved (localStorage) preference into this on
/// startup and on every toggle change. Defaults to on.
pub struct AutoCheckEnabled(pub AtomicBool);
impl Default for AutoCheckEnabled {
    fn default() -> Self {
        Self(AtomicBool::new(true))
    }
}

/// Pushed to the frontend as the "update-status" event. `state` is the
/// discriminant JS switches on; the extra fields are only present on
/// the variants that need them.
#[derive(Clone, Serialize)]
#[serde(tag = "state")]
pub enum UpdateStatus {
    #[serde(rename = "checking")]
    Checking,
    #[serde(rename = "available")]
    Available { version: String },
    #[serde(rename = "downloading")]
    Downloading { percent: u8 },
    #[serde(rename = "up-to-date")]
    UpToDate { version: String },
    #[serde(rename = "error")]
    Error { message: String },
}

fn emit_status(app: &AppHandle, status: UpdateStatus) {
    let _ = app.emit("update-status", status);
}

/// Shared by the manual "Check for Updates Now" button and the
/// background loop, so both behave identically and the UI only ever
/// has to listen for one event stream.
async fn run_check(app: &AppHandle) {
    emit_status(app, UpdateStatus::Checking);

    let updater = match app.updater() {
        Ok(u) => u,
        Err(err) => {
            eprintln!("Updater unavailable: {err}");
            emit_status(
                app,
                UpdateStatus::Error {
                    message: "Couldn't check for updates. Check your internet connection and try again.".into(),
                },
            );
            return;
        }
    };

    match updater.check().await {
        Ok(Some(update)) => {
            let version = update.version.clone();
            if let Some(pending) = app.try_state::<PendingUpdate>() {
                *pending.0.lock().unwrap() = Some(update);
            }
            emit_status(app, UpdateStatus::Available { version });
        }
        Ok(None) => {
            let version = app.package_info().version.to_string();
            emit_status(app, UpdateStatus::UpToDate { version });
        }
        Err(err) => {
            // Log the real cause for debugging but keep the message the
            // user sees generic and non-alarming — same call Playnck's
            // main.js makes about not forwarding raw error text to the
            // renderer.
            eprintln!("Update check failed: {err}");
            emit_status(
                app,
                UpdateStatus::Error {
                    message: "Couldn't check for updates. Check your internet connection and try again.".into(),
                },
            );
        }
    }
}

/// Runs once, 45 minutes apart, for as long as the app process is
/// alive. Skips the actual check in dev builds (`npm run tauri dev`)
/// and whenever the user has turned the Settings toggle off.
pub fn spawn_background_checks(app: AppHandle) {
    std::thread::spawn(move || {
        // Small delay so this can't possibly fire before updater.js has
        // registered its "update-status" listener on first launch.
        std::thread::sleep(Duration::from_secs(5));
        loop {
            let should_check = app
                .try_state::<AutoCheckEnabled>()
                .map(|s| s.0.load(Ordering::Relaxed))
                .unwrap_or(true);

            if should_check && !cfg!(debug_assertions) {
                tauri::async_runtime::block_on(run_check(&app));
            }

            std::thread::sleep(Duration::from_secs(45 * 60));
        }
    });
}

#[derive(Serialize)]
pub struct CheckStarted {
    pub started: bool,
    pub reason: Option<String>,
}

/// Manual "Check for Updates Now" button.
#[tauri::command]
pub async fn check_for_updates(app: AppHandle) -> CheckStarted {
    if cfg!(debug_assertions) {
        return CheckStarted {
            started: false,
            reason: Some("Updates only run in the installed app, not in `npm run tauri dev`.".into()),
        };
    }
    run_check(&app).await;
    CheckStarted {
        started: true,
        reason: None,
    }
}

/// "Download & Install" button — only enabled once a check has left an
/// `Update` sitting in `PendingUpdate`.
///
/// Note: on Windows, Tauri exits the process partway through
/// `download_and_install` in order to hand off to the NSIS installer — a
/// limitation of Windows installers, not something this code can avoid
/// (see Tauri's updater docs). Tauri's own updater→NSIS integration is
/// what relaunches Chronon once that silent install finishes; the
/// `app.restart()` call at the end of this function is the equivalent
/// explicit relaunch for any platform where the process *doesn't* exit
/// during the call above (this matches Tauri's own official example for
/// this command).
#[tauri::command]
pub async fn install_update_now(app: AppHandle, pending: State<'_, PendingUpdate>) -> Result<(), String> {
    let update = pending.0.lock().unwrap().take();
    let Some(update) = update else {
        return Err("No update is ready to install yet. Check for updates first.".into());
    };

    let downloaded = Arc::new(AtomicU64::new(0));
    let total = Arc::new(AtomicU64::new(0));
    let progress_app = app.clone();
    let progress_downloaded = downloaded.clone();
    let progress_total = total.clone();

    let result = update
        .download_and_install(
            move |chunk_length, content_length| {
                if let Some(len) = content_length {
                    progress_total.store(len, Ordering::Relaxed);
                }
                let now = progress_downloaded.fetch_add(chunk_length as u64, Ordering::Relaxed) + chunk_length as u64;
                let known_total = progress_total.load(Ordering::Relaxed);
                let percent = if known_total > 0 {
                    ((now as f64 / known_total as f64) * 100.0).min(100.0) as u8
                } else {
                    0
                };
                emit_status(&progress_app, UpdateStatus::Downloading { percent });
            },
            || {
                // Download finished; Tauri moves straight into installing
                // as soon as this returns. On Windows the process exits
                // here, so nothing after `.await` below runs in practice
                // on that platform.
            },
        )
        .await;

    if let Err(err) = result {
        eprintln!("Update install failed: {err}");
        emit_status(
            &app,
            UpdateStatus::Error {
                message: "Couldn't download or install the update. Check your internet connection and try again.".into(),
            },
        );
        return Err("Couldn't download or install the update.".into());
    }

    // Only reached on platforms where installing doesn't already exit
    // the process for us (Windows exits during the call above).
    app.restart();
    #[allow(unreachable_code)]
    Ok(())
}

#[tauri::command]
pub fn get_app_version(app: AppHandle) -> String {
    app.package_info().version.to_string()
}

#[tauri::command]
pub fn set_auto_check_updates(enabled: bool, state: State<'_, AutoCheckEnabled>) {
    state.0.store(enabled, Ordering::Relaxed);
}
